apps/smart-contracts/core/helpers/uniswap.ts

```diff
@@ -0,0 +1,55 @@
+import { BigNumber } from 'ethers'
+import JSBI from 'jsbi'
+import { parseEther } from '@ethersproject/units'
+import { encodeSqrtRatioX96, nearestUsableTick, TickMath, TICK_SPACINGS } from '@uniswap/v3-sdk'
+import { POOL_FEE_TIER } from 'prepo-constants'
+
+/**
+ * In Uniswap V3, the price is stored as a square root because of the geometric nature of the core
+ * AMM algorithm, x*y=k. The X96 stands for 96 bits reserved for the value after the decimal point.
+ * sqrtPriceX96 is stored as a uint160, meaning that there are 160 - 96 = 64 integer bits.
+ *
+ * Wei price = sqrtPriceX96 ** 2 / (2 ** 192), derivation taken from
+ * https://docs.uniswap.org/sdk/guides/fetching-prices.
+ */
+export function getWeiFromSqrtX96(sqrtX96Price: BigNumber): BigNumber {
+  return sqrtX96Price.pow(2).mul(parseEther('1')).div(BigNumber.from(2).pow(192))
+}
+
+/**
+ * Takes in a SqrtPriceX96 returned from Uniswap's SDK (which uses a different BigNumber
+ * library, JSBI) and converts it to the nearest tick.
+ *
+ * The desired price range passed in might not exactly sit within the pool's precision,
+ * so we must find the nearest tick rather than directly converting to it. Liquidity
+ * provision will revert if an invalid tick is provided (e.g. Specify a price of $0.50,
+ * but AMM only has 0.499 and 0.501 as valid prices).
```

Comment:
```suggestion
 * The converted tick might not be a valid tick for the pool,
 * so we must find the nearest tick rather than directly converting to it.
```
and move this above the `nearestUsableTick()` call

and then add this to the top of the file (since it applies to all functions):
```
 * Liquidity provision will revert if an invalid tick is provided, which is why the nearest valid tick/price must be found/used for the below functions.
 * E.g. If a pool had subsequent valid tick prices of $0.499 and $0.501, then $0.50 would be an invalid tick price.
```